### PR TITLE
rgw: adding request context structure

### PR DESCRIPTION
**Pull request by yuvalif**
this structure should be created at the frontend and trickle all the way to the RADOS layer. holding: dout prefix, optional yield and trace. in this commit, it was only added to the "complete()" interface.

in the future, it should be added to more sal interfaces, replacing the current way where dpp and optional yield are passed as sepearte arguments to all functions.
in addition, if more information would be needed, it should be possiblr to add that information to the request context struct without changing many function prototypes

basic test instructions:
https://gist.github.com/yuvalif/1c7f1e80126bed5fa79345efb27fe1b1

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [x] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/rgw/rgw_req_context.h [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+namespace opentelemetry::trace {
+  class Span;
+}
+
+class DoutPrefixProvider;
+
+// this struct holds information which is created at the frontend
+// and should trickle down through all function calls to the backend
+struct req_context {
+  const DoutPrefixProvider* const dpp{nullptr};
```
Comment (dang) [inline src/rgw/rgw_req_context.h:14]: Do we want a way to stack dpp?  Or are we going to use a single dpp call down the stack?  I know the top is generally a RGWOp instance, but is this overridden in places?

Comment (cbodley) [reply src/rgw/rgw_req_context.h:14]: in general, it doesn't make much sense for a struct to have all-const members. if you want all-const, then your `req_context` variables should themselves be const. this could just a raw struct with all of the constructors/operators left defaulted

passing `req_context` around as a const-ref seems like the right approach. if a function wants to override parts of the context, it can construct a new variable with updated pointers:
```c++
void foo2(const req_context& rctx);

void foo(const req_context& rctx)
{
  // wrap the caller's dpp
  CustomDoutPrefix dpp{*rctx.dpp};
  // add a child span to the trace
  auto span = tracer.add_span("child", rctx.span);

  foo2({.dpp = &dpp, .y = rctx.y, .span = span.get()});
}
```

File: src/rgw/rgw_req_context.h [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  const optional_yield y{null_yield};
+  opentelemetry::trace::Span* const span{nullptr};
```
Comment (cbodley) [inline src/rgw/rgw_req_context.h:16]: given a raw pointer to `Span`, it doesn't look like our `src/common/tracer.h` interfaces would let us add a child span, because `Tracer::add_span()` takes the parent span by shared_ptr

File: src/rgw/rgw_sal.h [hunk 3]
```diff
@@ -1430,7 +1431,7 @@ class Writer : public ObjectProcessor {
                        const char *if_match, const char *if_nomatch,
                        const std::string *user_data,
                        rgw_zone_set *zones_trace, bool *canceled,
-                       optional_yield y) = 0;
+                       const req_context& rctx) = 0;
```
Comment (cbodley) [inline src/rgw/rgw_sal.h:1434]: these sal changes also need to be applied to the overrides in driver/daos, driver/motr and driver/posix

Comment (yuvalif) [reply src/rgw/rgw_sal.h:1434]: ok. how do i control their compilation?
this would make sure i don't miss anything

File: src/rgw/rgw_req_context.h (hunk 1 above)
Comment (yuvalif) [reply src/rgw/rgw_req_context.h:14]: > in general, it doesn't make much sense for a struct to have all-const members. 

according to your suggestion, in order to change anything with the context, you need to create a new context with thew new variables in it, and copy over what remains the same.
this means, that there are no changes to the struct once created. in this case, why not make them all const?

> if you want all-const, then your `req_context` variables should themselves be const. 

this is what i did, but making the data members const assures that even if the user does not make the `req_context` variable const

> this could just a raw struct with all of the constructors/operators left defaulted

why is this better?

Comment (cbodley) [reply src/rgw/rgw_req_context.h:14]: > according to your suggestion, in order to change anything with the context, you need to create a new context with thew new variables in it, and copy over what remains the same. this means, that there are no changes to the struct once created. in this case, why not make them all const?

because the const vs. mutable decision should be up to the caller instead of hard-coded into the type itself. what if a function wants to start a child span in the middle? there's nothing wrong with using a mutable context in a local scope:

```c++
void foo2(const req_context& rctx);
void foo3(const req_context& rctx);

void foo(const req_context& rctx)
{
  req_context rctx2{rctx}; // normal copy ctor
  foo2(rctx2);

  // start a child span just for foo3()
  auto span = tracer.add_span("child", rctx.span);
  rctx2.span = span.get();

  foo3(rctx2);
}
```

> this is what i did, but making the data members const assures that even if the user does not make the `req_context` variable const

the chosen convention is to pass by const-ref, because we don't want a called function to change the context of its parent function. this kind of const-correctness should already be part of our code reviews

> > this could just a raw struct with all of the constructors/operators left defaulted
> 
> why is this better?

a raw struct with normal value semantics is easier to reason about and less code to maintain. adding const to members disables operations like move and copy assignment. this can be useful for classes that manage a resource, or otherwise rely on encapsulation to maintain invariants. but `req_context` is just a bundle of related variables. let's keep it simple!

File: src/rgw/rgw_sal.h (hunk 3 above)
Comment (cbodley) [reply src/rgw/rgw_sal.h:1434]: WITH_RADOSGW_POSIX is enabled by default. you could enable WITH_RADOSGW_MOTR and WITH_RADOSGW_DAOS, but you'd have to provide their dependencies manually. i don't think it's really worth trying to compile those, especially if your interface changes are trivial like this

File: src/rgw/rgw_req_context.h [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+#include "common/tracer.h"
+
+class DoutPrefixProvider;
+
+// this struct holds information which is created at the frontend
+// and should trickle down through all function calls to the backend
+struct req_context {
+  const DoutPrefixProvider* dpp{nullptr};
+  const optional_yield y{null_yield};
+  jspan span;
```
Comment (cbodley) [inline src/rgw/rgw_req_context.h:14]: could we change `add_span()` or add an overload that takes the raw `Span` pointer instead? storing a shared_ptr in `req_context` means that we have to do atomic refs every time we copy or construct a new one. using a raw pointer leaves ownership to the stack where it belongs

File: src/rgw/rgw_req_context.h [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+#include "common/async/yield_context.h"
+#include "common/tracer.h"
+
+class DoutPrefixProvider;
+
+// this struct holds information which is created at the frontend
+// and should trickle down through all function calls to the backend
+struct req_context {
+  const DoutPrefixProvider* dpp{nullptr};
+  const optional_yield y{null_yield};
```
[the remark this answers is not in the document]
Comment (cbodley) [reply src/rgw/rgw_req_context.h:13]: there are ops like [RGWDeleteMultiObj](https://github.com/ceph/ceph/blob/d6b8162/src/rgw/rgw_op.cc#L7202-L7203) that spawn child coroutines and override the yield context, so i think it's fine for the `optional_yield` to be mutable as well

File: src/rgw/rgw_req_context.h (hunk 2 above)
Comment (yuvalif) [reply src/rgw/rgw_req_context.h]: made it a raw pointer again.
no need for a new `Tracer::add_span()` API, as I'm just using the `SpanContext` API.
see: https://github.com/ceph/ceph/pull/53547/files#diff-d07015a33dfd523c87c32b6eef5fb7b564abe0e648875c7fea21922d21565020R496

File: src/rgw/rgw_req_context.h (hunk 4 above)
Comment (yuvalif) [reply src/rgw/rgw_req_context.h]: see: https://github.com/ceph/ceph/pull/53547#discussion_r1332841614

File: src/rgw/rgw_sal.h (hunk 3 above)
Comment (dang) [reply src/rgw/rgw_sal.h:1434]: MOTR and DAOS break regularly.  I think I'm the only one who builds them.

**Comment by yuvalif** [comment]
jenkins test api

File: src/rgw/driver/rados/rgw_putobj_processor.cc [hunk 6]
```diff
@@ -341,8 +341,14 @@ int AtomicObjectProcessor::complete(size_t accounted_size,
...
                                     rgw_zone_set *zones_trace,
-                                    bool *pcanceled, optional_yield y)
+                                    bool *pcanceled, 
+                                    const req_context& rctx)
 {
+  jspan span;
+  if (rctx.span) {
+    span = tracing::rgw::tracer.add_span("complete atomic operation", rctx.span->GetContext());
+  }
+
```
Comment (cbodley) [inline src/rgw/driver/rados/rgw_putobj_processor.cc:351]: these new spans aren't really part of the request context feature, can we move these to a separate pr?

it seems like we could emit a single span in `RGWRados::Object::Write::_do_write_meta()` to cover just the index transaction instead of emitting 3 different kinds of spans at this layer. the parent spans should know whether they're multipart or append operations

Comment (yuvalif) [reply src/rgw/driver/rados/rgw_putobj_processor.cc:351]: > these new spans aren't really part of the request context feature, can we move these to a separate pr?
> 

ok. used it here mainly to verify that I'm passing the right value in the request context.
will move to a separate PR.

> it seems like we could emit a single span in `RGWRados::Object::Write::_do_write_meta()` to cover just the index transaction instead of emitting 3 different kinds of spans at this layer. the parent spans should know whether they're multipart or append operations

ok. will do that in the other PR

Comment (yuvalif) [reply src/rgw/driver/rados/rgw_putobj_processor.cc:351]: @cbodley just to make sure that I understand the "complete()" API to which the `req_context` was added is invoked from `fetch_remote_obj()` not from `_do_write_meta()`.
I would add that to `_do_write_meta()` to write the span to the bi log, but i still need that in `fetch_remote_obj()` on the syncing side?

Comment (cbodley) [reply src/rgw/driver/rados/rgw_putobj_processor.cc:351]: > the "complete()" API to which the `req_context` was added is invoked from `fetch_remote_obj()` not from `_do_write_meta()`

each of these `*ObjectProcessor::complete()` functions are calling `obj_op.write_meta()` -> `_do_write_meta()`. you'd just need to pass in the `req_context` or span

Comment (yuvalif) [reply src/rgw/driver/rados/rgw_putobj_processor.cc:351]: ok. got it

File: src/rgw/rgw_op.cc [hunk 7]
```diff
@@ -4376,10 +4376,11 @@ void RGWPutObj::execute(optional_yield y)
   }
 
   tracepoint(rgw_op, processor_complete_enter, s->req_id.c_str());
+  const req_context rctx{this, s->yield, s->trace.get()};
```
Comment (cbodley) [inline src/rgw/rgw_op.cc:4379]: this doesn't compile when `WITH_JAEGER=OFF`:
```
src/rgw/rgw_op.cc: In member function ‘virtual void RGWPutObj::execute(optional_yield)’:
src/rgw/rgw_op.cc:4379:51: error: ‘class jspan’ has no member named ‘get’
 4379 |   const req_context rctx{this, s->yield, s->trace.get()};
      |                                                   ^~~
```

File: src/rgw/rgw_req_context.h [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+  class Span;
+}
+
+class DoutPrefixProvider;
+
+// this struct holds information which is created at the frontend
+// and should trickle down through all function calls to the backend
+struct req_context {
+  const DoutPrefixProvider* dpp{nullptr};
+  optional_yield y{null_yield};
```
Comment (cbodley) [inline src/rgw/rgw_req_context.h:15]: hmm, we've tried hard to make all uses of `null_yield` explicit so we can find and fix them all with grep. defaulting to `null_yield` here makes it possible for a code path to be synchronous without any mention of `null_yield` in its stack

what do you think about removing this default? i'd expect that to disable `req_context`'s default ctor (because `optional_yield` doesn't have a default ctor), so should require callers to pass one into the aggregate initializer like `req_context{.y = null_yield}` or `const req_context rctx{this, s->yield, s->trace.get()}`

File: src/rgw/rgw_op.cc (hunk 7 above)
Comment (yuvalif) [reply src/rgw/rgw_op.cc:4379]: wonder if we need the full `std::unique_ptr` interface (e.g. `reset()`, `swap()`)?

Comment (yuvalif) [reply src/rgw/rgw_op.cc:4379]: fixed, now it compiles with `WITH_JAEGER=OFF`

**Comment by yuvalif** [comment]
jenkins test api

File: src/common/tracer.h [hunk 9]
```diff
@@ -67,7 +67,8 @@ struct jspan_context {
   jspan_context(bool sampled_flag, bool is_remote) {}
 };
 
-struct span_stub {
+namespace opentelemetry::trace {
+struct Span {
```
[the remark this answers is not in the document]
Comment (yuvalif) [reply src/common/tracer.h:71]: because there are places in the code where we use `opentelemetry::trace::Span`

[the remark this answers is not in the document]
Comment (cbodley) [reply src/common/tracer.h:71]: where? why are those not guarded by WITH_JAEGER? these wrappers in common/tracer.h are meant to hide/abstract the use of opentelemetry

[the remark this answers is not in the document]
Comment (yuvalif) [reply src/common/tracer.h:71]: the idea behind creating these dummy classes in `commin/tracer.h` is that the rest of the code is not littered with the `WITH_JAEGER` macro.
especially for the cases where function prototypes depends on that macro.

[the remark this answers is not in the document]
Comment (cbodley) [reply src/common/tracer.h:71]: we currently have a `jspan = shared_ptr<opentelemetry::trace::Span>` typedef. if code outside of common/tracer needs to refer directly to the `opentelemetry::trace::Span` type, then we should add a separate typedef for that instead of adding things to another library's namespace. for example, maybe `jspan = opentelemetry::trace::Span` and `jspan_ptr = shared_ptr<Span>`?

[the remark this answers is not in the document]
Comment (cbodley) [reply src/common/tracer.h:71]: that said, i couldn't find anything outside of common/tracer that refers to the `Span` type directly. if another PR is adding those, we can move this discussion there

[the remark this answers is not in the document]
Comment (yuvalif) [reply src/common/tracer.h:71]: I will add the jspan_ptr.
This pr is using the Span typw directly, as it is stored as a bare pointer in the req_context struct

[the remark this answers is not in the document]
Comment (cbodley) [reply src/common/tracer.h:71]: oops, no wonder my grep missed it :)

[the remark this answers is not in the document]
Comment (yuvalif) [reply src/common/tracer.h]: fixed here: 0f34499f85ef519f3a64f2ba6efe4755a1098ee6

File: src/common/tracer.h [hunk 10]
```diff
@@ -79,26 +80,16 @@ struct span_stub {
...
- public:
-  span_stub& operator*() { return span; }
-  const span_stub& operator*() const { return span; }
-
-  span_stub* operator->() { return &span; }
-  const span_stub* operator->() const { return &span; }
-
-  operator bool() const { return false; }
-};
+using jspan_ptr = std::shared_ptr<jspan>;
```
Comment (cbodley) [inline src/common/tracer.h:83]: these wrappers avoided shared_ptr before. can we just rename jspan->jspan_ptr and span_stub->jspan?

Comment (cbodley) [reply src/common/tracer.h:83]: when `WITH_JAEGER=ON` but tracing is disabled, isn't the shared_ptr always initialized with a 'noop span' rather than empty?

if this stub has different semantics, i worry that code that is tested and working when WITH_JAEGER=ON will crash with null pointer exceptions when built with OFF

**Comment by yuvalif** [comment]
jenkins test make check

File: src/common/tracer.h [hunk 11]
```diff
@@ -79,26 +80,25 @@ struct span_stub {
...
-  span_stub span;
- public:
-  span_stub& operator*() { return span; }
-  const span_stub& operator*() const { return span; }
-
-  span_stub* operator->() { return &span; }
-  const span_stub* operator->() const { return &span; }
-
+struct jspan_ptr {           
+  jspan span;
```
Comment (cbodley) [inline src/common/tracer.h:84]: is there a reason this needs to be public now?

Comment (cbodley) [reply src/common/tracer.h:84]: i think it should remain private if there's no reason to change it. but this doesn't need to delay qa

**Comment by yuvalif** [comment]
teuthology results: http://pulpito.front.sepia.ceph.com/yuvalif-2023-10-12_08:15:46-rgw-wip-yuval-req-context-distro-default-smithi/

* multisite failures (no crashes)
* known test issue: `DEBIAN_FRONTEND=noninteractive sudo -E apt-get -y --force-yes install python-dev`
* known ragweed failure: `r_test_multipart_index_versioning.test`
* failure to start the rabbitmq service (unrelated to the PR, not sure if it is a known issue or not):
```
sudo chkconfig rabbitmq-server on
sudo: chkconfig: command not found
```

**Comment by cbodley** [comment]
> failure to start the rabbitmq service (unrelated to the PR, not sure if it is a known issue or not):

that looks like a recent regression from https://github.com/ceph/ceph/pull/53517 which bumped the `rgw/notifications` suite from centos8 to 9. i opened https://tracker.ceph.com/issues/63205 to track this; would you be willing to look into it?

**Comment by cbodley** [comment]
`qa/tasks/rabbitmq.py` also uses `yum` commands that will only work on centos. ideally we'd be able to run these tests on all supported distros
